dbcon/execplan/treenode.h

```diff
@@ -1031,41 +1034,75 @@ inline IDB_Decimal TreeNode::getDecimalVal()
         case CalpontSystemCatalog::INT:
         case CalpontSystemCatalog::SMALLINT:
         case CalpontSystemCatalog::TINYINT:
-            fResult.decimalVal.value = (int64_t)(fResult.intVal * pow((double)10.0, fResultType.scale));
-            fResult.decimalVal.scale = fResultType.scale;
-            fResult.decimalVal.precision = fResultType.precision;
+            fResult.decimalVal = IDB_Decimal(
+                                     (int64_t)(fResult.intVal * pow((double)10.0, fResultType.scale)),
+                                     fResultType.scale, fResultType.precision);
             break;
 
         case CalpontSystemCatalog::UBIGINT:
+        {
+            int64_t temp = (int64_t)(fResult.uintVal * pow((double)10.0, fResultType.scale));
+
+            if (fResultType.isWideDecimalPrecision())
+                fResult.decimalVal = IDB_Decimal(0, fResultType.scale,
+                                         fResultType.precision, (int128_t)temp);
+            else
+                fResult.decimalVal = IDB_Decimal(temp, fResultType.scale,
+                                         fResultType.precision);
+
+            break;
+        }
+
         case CalpontSystemCatalog::UMEDINT:
         case CalpontSystemCatalog::UINT:
         case CalpontSystemCatalog::USMALLINT:
         case CalpontSystemCatalog::UTINYINT:
-            fResult.decimalVal.value = (int64_t)(fResult.uintVal * pow((double)10.0, fResultType.scale));
-            fResult.decimalVal.scale = fResultType.scale;
-            fResult.decimalVal.precision = fResultType.precision;
+            fResult.decimalVal = IDB_Decimal(
+                                     (int64_t)(fResult.uintVal * pow((double)10.0, fResultType.scale)),
+                                     fResultType.scale, fResultType.precision);
             break;
 
         case CalpontSystemCatalog::LONGDOUBLE:
+        {
+            long double dlScaled = fResult.longDoubleVal;
+
+            if (fResultType.scale > 0)
             {
-                long double dlScaled = fResult.longDoubleVal;
-                if (fResultType.scale > 0)
-                {
-                    dlScaled= fResult.longDoubleVal * pow((double)10.0, fResultType.scale);
-                }
-                if (dlScaled > (double)MAX_BIGINT)
-                {
-                    throw logging::InvalidConversionExcept("TreeNode::getDecimalVal: decimal overflow.");
-                }
-                fResult.decimalVal.value = (int64_t)roundl((fResult.longDoubleVal * pow((double)10.0, fResultType.scale)));
-                fResult.decimalVal.scale = fResultType.scale;
-                fResult.decimalVal.precision = fResultType.precision;
+                dlScaled= fResult.longDoubleVal * pow((double)10.0, fResultType.scale);
             }
+            if (dlScaled > (double)MAX_BIGINT)
+            {
+                throw logging::InvalidConversionExcept("TreeNode::getDecimalVal: decimal overflow.");
+            }
+
```

Comment:
Is the above `if` block still correct with wide decimal support added? I think it should now check for max int128_t value.

The below code could probably do:
- if dlScaled is greater than max int128_t, then error.
- if dlScaled is greated than MAX_BIGINT, the go with the wide-decimal constructor.
- otherwise, go with the narrow decimal constructor.

Reply:
Good catch, the original `if` block was incorrect. I am now checking if `dlScaled` is in the range of an `int64_t`. If it is, then I call the narrow decimal ctor. If it is outside the range of `int64_t`, I am silently saturating to `int128_t` and calling the wide decimal ctor. I think raising the exception is also an option here, instead of saturating.